src/common-subexpr-elim.rkt

```diff
@@ -29,43 +29,38 @@
 (define (add-name! name)
   (set-add! (*names*) name))
 
-;;;;;; fuse let expressions
-
-(define (visit-let/fuse visitor let_ vars vals body #:ctx [ctx '()])
-  (match (list let_ vars body)
-   [(list 'let (list var) `(let ([,var2 ,val2]) ,body2))          ; (let ([a ...]) (let ([b ...]) ...))
-    (visit/ctx visitor
-              `(let* (,(list var (car vals)) ,(list var2 val2)) ,body2)
-               ctx)]
-   [(list 'let (list var) `(let* ([,vars2 ,vals2] ...) ,body2))   ; (let ([a ...]) (let* ([b ...] ...) ...))
-    (visit/ctx visitor
-              `(let* (,@(map list (cons var vars2) (cons (car vals) vals2))) ,body2)
-               ctx)]
-   [(list 'let* _ `(let ([,var2 ,val2]) ,body2))   ; (let* ([a ...] ...) (let ([b ...]) ...))
-    (visit/ctx visitor
-              `(let* (,@(map list vars vals) ,(list var2 val2)) ,body2)
-               ctx)]
-   [(list 'let* _ `(let* ([,vars2 ,vals2] ...) ,body2))   ; (let* ([a ...] ...) (let* ([b ...] ...) ...))
-    (visit/ctx visitor
-              `(let* (,@(map list (append vars vars2) (append vals vals2))) ,body2)
-               ctx)]
-   [_
-    `(,let_ (,@(for/list ([var vars] [val vals]) (list var (visit/ctx visitor val ctx))))
-          ,(visit/ctx visitor body ctx))]))
+(define (clear-names!)
+  (set-clear! (*names*)))
 
-(define/transform-expr (fuse-let expr)
-  [visit-let_ visit-let/fuse])
-
-;;;;;; main cse
-
-(define (syntax? name)
-  (set-member? '(if let let* while while* for for* tensor tensor* !) name))
+;; fuse let expressions
 
-(define (filter-no-edges deps)
-  (map car (filter (compose null? cdr) deps)))
+(define/transform-expr (fuse-let expr)
```

Comment:
Ok, it took a while to convince me this was correct, but it is. I wonder if it would be simpler to write in two passes. First, convert all `let`s with a single variable in them to a `let*`s. Then fuse `let*`s but not `let`s.

Also, perhaps we should offer let-fusion from `transform.rkt`?

Reply:
That sounds like a much better design.